File: lib/domain/character/character.dart
import '../combat/combat.dart';
import '../economy/inventory.dart';
import '../economy/item.dart';
import '../survival/survival_tables.dart';
import '../world/coords.dart';
import 'attributes.dart';

/// Estado de um personagem. É o que a campanha persiste — junto com o layout do
/// mundo, é tudo que precisa sair para disco.
class Character {
  Character({
    required this.id,
    required this.name,
    required this.attributes,
    required this.position,
    required this.homeSettlementId,
    this.level = CitizenLevel.survivor,
    this.credits = 250,
    int? hp,
    this.hunger = SurvivalTables.maxVital,
    this.thirst = SurvivalTables.maxVital,
    this.energy = 10,
    this.rerollsUsed = 0,
    this.starvingStreak = 0,
    this.dead = false,
    this.deathReason,
    Inventory? inventory,
    this.travellingTo,
    this.travelDaysRemaining = 0,
    this.statusOffset = 0,
  })  : inventory = inventory ?? Inventory(),
        hp = hp ?? Combatant.maxHpFor(attributes);

  final String id;
  String name;

  AttributeSet attributes;

  /// Onde o personagem está no mundo.
  TileCoord position;

  /// Cidade de origem — para onde ele volta se morrer de forma não permanente.
  String homeSettlementId;

  CitizenLevel level;

  int credits;
  int hp;
  int hunger;
  int thirst;

  /// Horas de trabalho disponíveis no dia. Energéticos aumentam.
  int energy;

  int rerollsUsed;

  /// Resets consecutivos com Fome **e** Sede zeradas. Chegando ao limite do
  /// GDD, é morte permanente por abandono.
  int starvingStreak;

  bool dead;
  String? deathReason;

  final Inventory inventory;

  /// Se estiver viajando: destino e dias restantes. Enquanto viaja, o jogador
  /// fica bloqueado para ações — restrição explícita do canvas.
  String? travellingTo;
  int travelDaysRemaining;

  /// Ajuste de Status ganho ou perdido em jogo (combate, política, luxo).
  int statusOffset;

  bool get isTravelling => travelDaysRemaining > 0;

  bool get canAct => !dead && !isTravelling;

  int get maxHp => Combatant.maxHpFor(attributes);

  /// Status efetivo: o rolado na criação mais o que o jogo somou.
  int get effectiveStatus => attributes[Attribute.status] + statusOffset;

  bool get canReroll =>
      rerollsUsed < AttributeSet.maxRerolls && level == CitizenLevel.survivor;

  /// Consome um item, aplicando restauração e custos.
  ///
  /// Devolve `false` se o item não está no inventário — a UI usa isso para não
  /// permitir clique em item sem saldo.
  bool consume(ItemId id) {
    if (!inventory.remove(id, 1)) return false;
    final def = ItemCatalog.of(id);

    hunger = (hunger + def.restoresHunger - def.hungerCost)
        .clamp(SurvivalTables.minVital, SurvivalTables.maxVital);
    thirst = (thirst + def.restoresThirst - def.thirstCost)
        .clamp(SurvivalTables.minVital, SurvivalTables.maxVital);
    energy += def.energyBonus;
    statusOffset += def.statusBonus;
    return true;
  }

  /// Aplica o consumo do dia e devolve o que aconteceu.
  DayEndOutcome applyUpkeep(Upkeep upkeep) {
    hunger = (hunger - upkeep.hunger).clamp(SurvivalTables.minVital, SurvivalTables.maxVital);
    thirst = (thirst - upkeep.thirst).clamp(SurvivalTables.minVital, SurvivalTables.maxVital);

    final starving = hunger <= SurvivalTables.starvationThreshold;
    final dehydrated = thirst <= SurvivalTables.starvationThreshold;

    var damage = 0;
    if (starving) damage += SurvivalTables.starvationDamage;
    if (dehydrated) damage += SurvivalTables.starvationDamage;

    if (damage > 0) {
      hp = (hp - damage).clamp(0, maxHp);
    }

    if (starving && dehydrated) {
      starvingStreak++;
    } else {
      starvingStreak = 0;
    }

    // Morte permanente é reservada ao abandono: exige dias seguidos com as
    // duas barras zeradas, ou HP em zero por inanição.
    final permanent = starvingStreak >= SurvivalTables.consecutiveDaysToDeath ||
        (hp <= 0 && (starving || dehydrated));

    if (permanent) {
      dead = true;
      deathReason = 'Morte permanente por abandono (fome e sede).';
    }

    return DayEndOutcome(
      hungerLost: upkeep.hunger,
      thirstLost: upkeep.thirst,
      hpLost: damage,
      starving: starving,
      dehydrated: dehydrated,
      died: permanent,
    );
  }

  /// Requisitos de progressão do GDD (seção 5).
  bool meetsRequirementsFor(CitizenLevel target) => switch (target) {
        CitizenLevel.survivor => true,
        CitizenLevel.farmer => credits >= 1500,
        CitizenLevel.industrialist =>
          credits >= 12000 && inventory.estimatedValue >= 5000,
        CitizenLevel.elite =>
          credits >= 60000 && effectiveStatus >= 12,
      };

  bool promote() {
    final next = level.next;
    if (next == null) return false;
    if (!meetsRequirementsFor(next)) return false;
    level = next;
    return true;
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'attributes': attributes.toJson(),
        'position': position.toJson(),
        'homeSettlementId': homeSettlementId,
        'level': level.name,
        'credits': credits,
        'hp': hp,
        'hunger': hunger,
        'thirst': thirst,
        'energy': energy,
        'rerollsUsed': rerollsUsed,
        'starvingStreak': starvingStreak,
        'dead': dead,
        'deathReason': deathReason,
        'inventory': inventory.toJson(),
        'travellingTo': travellingTo,
        'travelDaysRemaining': travelDaysRemaining,
        'statusOffset': statusOffset,
      };

  factory Character.fromJson(Map<String, dynamic> json) => Character(
        id: json['id'] as String,
        name: json['name'] as String,
        attributes:
            AttributeSet.fromJson(json['attributes'] as Map<String, dynamic>),
        position: TileCoord.fromJson(json['position'] as Map<String, dynamic>),
        homeSettlementId: json['homeSettlementId'] as String,
        level: CitizenLevel.values.byName(json['level'] as String),
        credits: (json['credits'] as num).toInt(),
        hp: (json['hp'] as num).toInt(),
        hunger: (json['hunger'] as num).toInt(),
        thirst: (json['thirst'] as num).toInt(),
        energy: (json['energy'] as num).toInt(),
        rerollsUsed: (json['rerollsUsed'] as num).toInt(),
        starvingStreak: (json['starvingStreak'] as num?)?.toInt() ?? 0,
        dead: json['dead'] as bool? ?? false,
        deathReason: json['deathReason'] as String?,
        inventory: Inventory.fromJson(json['inventory'] as Map<String, dynamic>),
        travellingTo: json['travellingTo'] as String?,
        travelDaysRemaining: (json['travelDaysRemaining'] as num?)?.toInt() ?? 0,
        statusOffset: (json['statusOffset'] as num?)?.toInt() ?? 0,
      );
}

/// O que aconteceu com o personagem no fechamento do dia.
class DayEndOutcome {
  const DayEndOutcome({
    required this.hungerLost,
    required this.thirstLost,
    required this.hpLost,
    required this.starving,
    required this.dehydrated,
    required this.died,
  });

  final int hungerLost;
  final int thirstLost;
  final int hpLost;
  final bool starving;
  final bool dehydrated;
  final bool died;
}

File: web3d/src/character/character.ts
import { maxHpFor } from '../combat/combat';
import { Inventory } from '../economy/inventory';
import { itemDef } from '../economy/item';
import { SurvivalTables, type Upkeep } from '../survival/survival';
import { TileCoord } from '../world/coords';
import {
  AttributeSet,
  MAX_REROLLS,
  nextCitizenLevel,
  type CitizenLevel,
} from './attributes';

/**
 * Estado de um personagem.
 *
 * Junto com o layout do mundo, é tudo que precisa sair para disco: o terreno é
 * função pura da seed e se regenera igual, então persistir chunk seria guardar
 * o que já se sabe calcular.
 *
 * ## Por que a morte permanente é tão difícil de alcançar
 *
 * O GDD reserva a morte permanente ao **abandono**, não ao azar. Perder uma
 * briga não mata; passar fome um dia não mata. Só matam dias seguidos com as
 * duas barras zeradas, ou HP em zero *por inanição*. Num jogo de reset diário,
 * uma morte que pode acontecer por descuido de um dia faria o jogador parar de
 * arriscar — e o risco é o jogo.
 */

export interface CharacterOptions {
  readonly id: string;
  readonly name: string;
  readonly attributes: AttributeSet;
  readonly position: TileCoord;
  readonly homeSettlementId: string;
  readonly level?: CitizenLevel;
  readonly credits?: number;
  readonly hp?: number;
  readonly hunger?: number;
  readonly thirst?: number;
  readonly energy?: number;
  readonly rerollsUsed?: number;
  readonly starvingStreak?: number;
  readonly dead?: boolean;
  readonly deathReason?: string | null;
  readonly inventory?: Inventory;
  readonly travellingTo?: string | null;
  readonly travelDaysRemaining?: number;
  readonly statusOffset?: number;
}

/** O que aconteceu com o personagem no fechamento do dia. */
export interface DayEndOutcome {
  readonly hungerLost: number;
  readonly thirstLost: number;
  readonly hpLost: number;
  readonly starving: boolean;
  readonly dehydrated: boolean;
  readonly died: boolean;
}

export class Character {
  readonly id: string;
  name: string;
  attributes: AttributeSet;
  /** Onde o personagem está no mundo. */
  position: TileCoord;
  /** Cidade de origem — para onde volta numa morte não permanente. */
  homeSettlementId: string;
  level: CitizenLevel;
  credits: number;
  hp: number;
  hunger: number;
  thirst: number;
  /** Horas de trabalho disponíveis no dia. Energéticos aumentam. */
  energy: number;
  rerollsUsed: number;
  /** Resets consecutivos com Fome **e** Sede zeradas. */
  starvingStreak: number;
  dead: boolean;
  deathReason: string | null;
  readonly inventory: Inventory;
  /** Destino e dias restantes. Viajando, o jogador não age. */
  travellingTo: string | null;
  travelDaysRemaining: number;
  /** Status ganho ou perdido em jogo: combate, política, luxo. */
  statusOffset: number;

  constructor(options: CharacterOptions) {
    this.id = options.id;
    this.name = options.name;
    this.attributes = options.attributes;
    this.position = options.position;
    this.homeSettlementId = options.homeSettlementId;
    this.level = options.level ?? 'survivor';
    this.credits = options.credits ?? 250;
    this.hunger = options.hunger ?? SurvivalTables.maxVital;
    this.thirst = options.thirst ?? SurvivalTables.maxVital;
    this.energy = options.energy ?? 10;
    this.rerollsUsed = options.rerollsUsed ?? 0;
    this.starvingStreak = options.starvingStreak ?? 0;
    this.dead = options.dead ?? false;
    this.deathReason = options.deathReason ?? null;
    this.inventory = options.inventory ?? new Inventory();
    this.travellingTo = options.travellingTo ?? null;
    this.travelDaysRemaining = options.travelDaysRemaining ?? 0;
    this.statusOffset = options.statusOffset ?? 0;
    this.hp = options.hp ?? maxHpFor(options.attributes);
  }

  get isTravelling(): boolean {
    return this.travelDaysRemaining > 0;
  }

  get canAct(): boolean {
    return !this.dead && !this.isTravelling;
  }

  get maxHp(): number {
    return maxHpFor(this.attributes);
  }

  /** Status efetivo: o rolado na criação mais o que o jogo somou. */
  get effectiveStatus(): number {
    return this.attributes.get('status') + this.statusOffset;
  }

  get canReroll(): boolean {
    return this.rerollsUsed < MAX_REROLLS && this.level === 'survivor';
  }

  /**
   * Consome um item, aplicando restauração e custos.
   *
   * Devolve `false` quando o item não está no inventário — a interface usa isso
   * para não deixar clicar em item sem saldo.
   */
  consume(id: string): boolean {
    if (!this.inventory.remove(id, 1)) return false;
    const def = itemDef(id);

    this.hunger = clampVital(
      this.hunger + def.restoresHunger - def.hungerCost,
    );
    this.thirst = clampVital(this.thirst + def.restoresThirst - def.thirstCost);
    this.energy += def.energyBonus;
    this.statusOffset += def.statusBonus;
    return true;
  }

  /** Aplica o consumo do dia e devolve o que aconteceu. */
  applyUpkeep(upkeep: Upkeep): DayEndOutcome {
    this.hunger = clampVital(this.hunger - upkeep.hunger);
    this.thirst = clampVital(this.thirst - upkeep.thirst);

    const starving = this.hunger <= SurvivalTables.starvationThreshold;
    const dehydrated = this.thirst <= SurvivalTables.starvationThreshold;

    let damage = 0;
    if (starving) damage += SurvivalTables.starvationDamage;
    if (dehydrated) damage += SurvivalTables.starvationDamage;

    if (damage > 0) {
      this.hp = Math.min(this.maxHp, Math.max(0, this.hp - damage));
    }

    if (starving && dehydrated) {
      this.starvingStreak++;
    } else {
      this.starvingStreak = 0;
    }

    // Morte permanente é reservada ao abandono: dias seguidos com as duas
    // barras zeradas, ou HP em zero *por inanição*. HP zerado numa briga não
    // entra aqui — perder uma briga não mata, por decisão do GDD.
    const permanent =
      this.starvingStreak >= SurvivalTables.consecutiveDaysToDeath ||
      (this.hp <= 0 && (starving || dehydrated));

    if (permanent) {
      this.dead = true;
      this.deathReason = 'Morte permanente por abandono (fome e sede).';
    }

    return {
      hungerLost: upkeep.hunger,
      thirstLost: upkeep.thirst,
      hpLost: damage,
      starving,
      dehydrated,
      died: permanent,
    };
  }

  /** Requisitos de progressão do GDD, seção 5. */
  meetsRequirementsFor(target: CitizenLevel): boolean {
    switch (target) {
      case 'survivor':
        return true;
      case 'farmer':
        return this.credits >= 1500;
      case 'industrialist':
        return this.credits >= 12000 && this.inventory.estimatedValue >= 5000;
      case 'elite':
        return this.credits >= 60000 && this.effectiveStatus >= 12;
    }
  }

  promote(): boolean {
    const next = nextCitizenLevel(this.level);
    if (next === null) return false;
    if (!this.meetsRequirementsFor(next)) return false;
    this.level = next;
    return true;
  }

  toJson(): Record<string, unknown> {
    return {
      id: this.id,
      name: this.name,
      attributes: this.attributes.toJson(),
      position: this.position.toJson(),
      homeSettlementId: this.homeSettlementId,
      level: this.level,
      credits: this.credits,
      hp: this.hp,
      hunger: this.hunger,
      thirst: this.thirst,
      energy: this.energy,
      rerollsUsed: this.rerollsUsed,
      starvingStreak: this.starvingStreak,
      dead: this.dead,
      deathReason: this.deathReason,
      inventory: this.inventory.toJson(),
      travellingTo: this.travellingTo,
      travelDaysRemaining: this.travelDaysRemaining,
      statusOffset: this.statusOffset,
    };
  }

  static fromJson(json: Record<string, unknown>): Character {
    const inteiro = (v: unknown, padrao: number): number => {
      const bruto = Number(v);
      return Number.isFinite(bruto) ? Math.trunc(bruto) : padrao;
    };
    const attributes = AttributeSet.fromJson(
      (json.attributes ?? {}) as Record<string, unknown>,
    );

    return new Character({
      id: String(json.id ?? ''),
      name: String(json.name ?? ''),
      attributes,
      position: TileCoord.fromJson(json.position as { x?: unknown; y?: unknown }),
      homeSettlementId: String(json.homeSettlementId ?? ''),
      level: (json.level as CitizenLevel) ?? 'survivor',
      credits: inteiro(json.credits, 250),
      hp: inteiro(json.hp, maxHpFor(attributes)),
      hunger: inteiro(json.hunger, SurvivalTables.maxVital),
      thirst: inteiro(json.thirst, SurvivalTables.maxVital),
      energy: inteiro(json.energy, 10),
      rerollsUsed: inteiro(json.rerollsUsed, 0),
      starvingStreak: inteiro(json.starvingStreak, 0),
      dead: json.dead === true,
      deathReason: (json.deathReason as string | null) ?? null,
      inventory: Inventory.fromJson(
        (json.inventory ?? {}) as Parameters<typeof Inventory.fromJson>[0],
      ),
      travellingTo: (json.travellingTo as string | null) ?? null,
      travelDaysRemaining: inteiro(json.travelDaysRemaining, 0),
      statusOffset: inteiro(json.statusOffset, 0),
    });
  }
}

/** Prende um vital entre o mínimo e o máximo da tabela. */
function clampVital(valor: number): number {
  return Math.min(
    SurvivalTables.maxVital,
    Math.max(SurvivalTables.minVital, valor),
  );
}
